Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp

```diff
@@ -3012,150 +3012,160 @@ RefPtr<StyleValue> Parser::parse_identifier_value(ComponentValue const& componen
+            a_val = params[3].percentage() / 100.0f;
 
-                u8 r = clamp(lroundf(r_val.percentage() * 2.55f), 0, 255);
-                u8 g = clamp(lroundf(g_val.percentage() * 2.55f), 0, 255);
-                u8 b = clamp(lroundf(b_val.percentage() * 2.55f), 0, 255);
-                return Color(r, g, b);
-            }
-        } else if (function.name().equals_ignoring_case("rgba")) {
-            if (params.size() != 4)
-                return {};
+        if (params[0].is(Token::Type::Dimension)
+            && params[1].is(Token::Type::Percentage)
+            && params[2].is(Token::Type::Percentage)) {
 
-            auto r_val = params[0];
-            auto g_val = params[1];
-            auto b_val = params[2];
-            auto a_val = params[3];
-
-            if (r_val.is(Token::Type::Number) && r_val.number().is_integer()
-                && g_val.is(Token::Type::Number) && g_val.number().is_integer()
-                && b_val.is(Token::Type::Number) && b_val.number().is_integer()
-                && a_val.is(Token::Type::Number)) {
-
-                auto r = r_val.to_integer();
-                auto g = g_val.to_integer();
-                auto b = b_val.to_integer();
-                auto a = clamp(lroundf(a_val.number_value() * 255.0f), 0, 255);
-                if (AK::is_within_range<u8>(r) && AK::is_within_range<u8>(g) && AK::is_within_range<u8>(b))
-                    return Color(r, g, b, a);
-
-            } else if (r_val.is(Token::Type::Percentage)
-                && g_val.is(Token::Type::Percentage)
-                && b_val.is(Token::Type::Percentage)
-                && a_val.is(Token::Type::Number)) {
-
-                auto r = r_val.percentage();
-                auto g = g_val.percentage();
-                auto b = b_val.percentage();
-                auto a = a_val.number_value();
-
-                u8 r_255 = clamp(lroundf(r * 2.55f), 0, 255);
-                u8 g_255 = clamp(lroundf(g * 2.55f), 0, 255);
-                u8 b_255 = clamp(lroundf(b * 2.55f), 0, 255);
-                u8 a_255 = clamp(lroundf(a * 255.0f), 0, 255);
-                return Color(r_255, g_255, b_255, a_255);
-            }
-        } else if (function.name().equals_ignoring_case("hsl")) {
-            if (params.size() != 3)
+            float numeric_value = params[0].dimension_value();
+            auto unit_string = params[0].dimension_unit();
+            auto angle_type = Angle::unit_from_name(unit_string);
+
+            if (!angle_type.has_value())
                 return {};
 
-            auto h_val = params[0];
-            auto s_val = params[1];
-            auto l_val = params[2];
+            auto angle = Angle { numeric_value, angle_type.release_value() };
 
-            if (h_val.is(Token::Type::Number)
-                && s_val.is(Token::Type::Percentage)
-                && l_val.is(Token::Type::Percentage)) {
+            float h_val = fmodf(angle.to_degrees(), 360.0f);
```

Comment:
According to https://www.w3.org/TR/css-color-4/#hsl-to-rgb the angle wrapping (and ensuring it's non-negative) should be handled inside `Color::from_hsla()`... which seems to be based on an older version of the spec, and also does the angle wrapping incorrectly.

It's up to you if you want to fix that now or leave it, I've made a note or myself to get to that at some point if you don't. :^)

Reply:
I can move the wrapping into `Color::from_hsla()`. I thought I would stay in LibWeb for this PR.

The CSS algorithm uses the range 0..100 for `s` and `l`, where `from_hsla` uses 0..1. Would you like to keep the existing interface or update to use the CSS interface?